# Part 1

## Explain the variables contained in the dataset. What do they refer to?

- **customerID:** A unique identifier for each customer.
- **gender:** The customer's gender.
- **SeniorCitizen:** Whether the customer is 65 years old or older.
- **Partner:** Whether the customer has a partner.
- **Dependents:** Whether the customer has dependents.
- **tenure:** The number of months the customer has been with the company.
- **PhoneService:** Whether the customer subscribes to phone service.
- **MultipleLines:** Whether the customer subscribes to multiple lines of phone service.
- **InternetService:** The type of internet service the customer subscribes to (DSL, fiber optic, or no internet service).
- **OnlineSecurity:** Whether the customer subscribes to online security service.
- **OnlineBackup:** Whether the customer subscribes to online backup service.
- **DeviceProtection:** Whether the customer subscribes to device protection service.
- **TechSupport:** Whether the customer subscribes to tech support service.
- **StreamingTV:** Whether the customer subscribes to streaming TV service.
- **StreamingMovies:** Whether the customer subscribes to streaming movies service.
- **Contract:** The type of contract the customer has (month-to-month, one year, or two years).
- **PaperlessBilling:** Whether the customer prefers paperless billing.
- **PaymentMethod:** The customer's payment method (electronic check, mailed check, or bank transfer).
- **MonthlyCharges:** The customer's monthly charges.
- **TotalCharges:** The customer's total charges over the lifetime of their account.
- **Churn:** Whether the customer churned (i.e., left the company) within the last month.

In [131]:
import pandas as pd

In [132]:
df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv')

In [133]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

- **In 'WA_Fn-UseC_-Telco-Customer-Churn' dataset, there are a total of 16 categorical variables. These categorical variables include: gender, SeniorCitizen, Partner, Dependents, PhoneService, MultipleLines, InternetService, OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV, StreamingMovies, Contract, PaperlessBilling, and PaymentMethod**

In [134]:
df.shape

(7043, 21)

- **The dimensions of the dataset (7043, 21) tells us that there are 7043 rows and 21 columns. This means that there are 7043 customers in the dataset, and each customer is described by 21 features.**

In [135]:
#Display all columns 
# Try changing the parameters:
with pd.option_context('display.max_rows', 21, 'display.max_columns', None):
    display(df)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  \
0              No  No phone service             DSL             No   
1             Yes                No             DSL            Yes   
2             Yes                No             DSL            Yes   
3              No  No phone service             DSL            Yes   
4             Yes                No     Fiber optic             No   
...           ...               ...             ...            ...   
7038          Yes               Yes             DSL            Yes   
7039          Yes               Yes     Fiber optic             No   
7040           No  No phone service             DSL            Yes   
7041          Yes               Yes     Fiber optic             No   
7042          Yes                No     Fiber optic            Yes   

     OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies  \
0             Yes               No          No          No              No   
1              No              Yes          No          No              No   
2             Yes               No          No          No              No   
3              No              Yes         Yes          No              No   
4              No               No          No          No              No   
...           ...              ...         ...         ...             ...   
7038           No              Yes         Yes         Yes             Yes   
7039          Yes              Yes          No         Yes             Yes   
7040           No               No          No          No              No   
7041           No               No          No          No              No   
7042           No              Yes         Yes         Yes             Yes   

            Contract PaperlessBilling              PaymentMethod  \
0     Month-to-month              Yes           Electronic check   
1           One year               No               Mailed check   
2     Month-to-month              Yes               Mailed check   
3           One year               No  Bank transfer (automatic)   
4     Month-to-month              Yes           Electronic check   
...              ...              ...                        ...   
7038        One year              Yes               Mailed check   
7039        One year              Yes    Credit card (automatic)   
7040  Month-to-month              Yes           Electronic check   
7041  Month-to-month              Yes               Mailed check   
7042        Two year              Yes  Bank transfer (automatic)   

      MonthlyCharges TotalCharges Churn  
0              29.85        29.85    No  
1              56.95       1889.5    No  
2              53.85       108.15   Yes  
3              42.30      1840.75    No  
4              70.70       151.65   Yes  
...              ...          ...   ...  
7038           84.80       1990.5    No  
7039          103.20       7362.9    No  
7040           29.60       346.45    No  
7041           74.40        306.6   Yes  
7042          105.65       6844.5    No  

[7043 rows x 21 columns]

In [136]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


- **The dataset appears to have a mix of categorical, numerical, and binary variables. It's important to note that the "TotalCharges" column might need to be converted to a numeric data type for proper analysis if it contains numerical values stored as strings.**
- **Based on the information, it seems that there are no missing values in the dataset. The "Non-Null Count" column for each variable indicates the number of non-null values for that variable. Since all the columns have the same count as the total number of entries, it suggests that there are no missing values in any of the columns.**

In [137]:
#Convert total charge to numeric data type
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

In [138]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [139]:
# Fill the NaN value in total charges with the mean value
df['TotalCharges'].fillna(df['TotalCharges'].mean(), inplace=True)

In [140]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [141]:
# Calculate churn rate
df['Churn'].value_counts(normalize=True) * 100

No     73.463013
Yes    26.536987
Name: Churn, dtype: float64

- The majority of customers did not churn. This means that the company is doing a good job.
- IT is important to note that the churn rate is still relatively high. This means that there is still room for improvement. The company can analyze the data further to identify the factors that are associated with churn and take steps to reduce it.

## How does tenure affect customer churn?

In [142]:
short_range = (0, 12)         # up to 12 months
intermediate_range = (13, 36)  # 13 to 36 months
long_range = (37, float('inf'))  # 37 months and above

#categorize tenure periods
def categorize_tenure(tenure):
    if tenure >= long_range[0]:
        return 'Long Period'
    elif tenure >= intermediate_range[0]:
        return 'Intermediate Period'
    elif tenure >= short_range[0]:
        return 'Short Period'
    else:
        return 'Unknown'
df['tenure_category'] = df['tenure'].apply(categorize_tenure)


In [143]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... TechSupport  \
0  No phone service             DSL             No  ...          No   
1                No             DSL            Yes  ...          No   
2                No             DSL            Yes  ...          No   
3  No phone service             DSL            Yes  ...         Yes   
4                No     Fiber optic             No  ...          No   

  StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No              No  Month-to-month              Yes   
1          No              No        One year               No   
2          No              No  Month-to-month              Yes   
3          No              No        One year               No   
4          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges TotalCharges  Churn  \
0           Electronic check          29.85        29.85     No   
1               Mailed check          56.95      1889.50     No   
2               Mailed check          53.85       108.15    Yes   
3  Bank transfer (automatic)          42.30      1840.75     No   
4           Electronic check          70.70       151.65    Yes   

       tenure_category  
0         Short Period  
1  Intermediate Period  
2         Short Period  
3          Long Period  
4         Short Period  

[5 rows x 22 columns]

In [144]:
df['tenure'].mean()

32.37114865824223

The average tenure of a customer is 32.37 months. This means that the average customer has been with the company for 32.37 months.

In [145]:
df.groupby('tenure_category')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

tenure_category Churn  percentage
0  Intermediate Period    No   74.461207
1  Intermediate Period   Yes   25.538793
2          Long Period    No   88.070643
3          Long Period   Yes   11.929357
4         Short Period    No   52.561757
5         Short Period   Yes   47.438243

- **For the "Intermediate Period," the churn percentage is around 25.54%, meaning that approximately 25.54% of customers in this category have churned.**
- **For the "Long Period," the churn percentage is lower at 11.93%.**
- **In the "Short Period" category, the churn percentage is relatively high at 47.44%.**
- **It seems that customers with longer tenures have lower churn percentages compared to those with shorter tenures.**
<br>
<br>
- **Customers in the "Long Period" category seem to have a higher likelihood of staying with the service, as indicated by the lower churn percentage.**
- **The "Short Period" category has a much higher churn percentage, suggesting that customers who have recently joined the service are more likely to churn.**
- **It might be valuable to investigate why customers with shorter tenures are churning at a higher rate. Are there any factors that are common among these customers?**

## Are customers with month-to-month contracts more likely to churn than customers with longer contracts?

In [146]:
df.groupby('Contract')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

Contract Churn  percentage
0  Month-to-month    No   57.290323
1  Month-to-month   Yes   42.709677
2        One year    No   88.730482
3        One year   Yes   11.269518
4        Two year    No   97.168142
5        Two year   Yes    2.831858

- **Customers with a "Month-to-month" contract have a churn percentage of about 42.71%. This suggests that almost 43% of customers with this contract type have churned.**
- **For customers with a "One year" contract, the churn percentage is significantly lower at 11.27%.**
- **Customers with a "Two year" contract exhibit an even lower churn percentage of only 2.83%.**
<br>
<br>
- **Customers with longer contract durations ("One year" and "Two year") seem to have much lower churn percentages compared to those with "Month-to-month" contracts.**
- **This indicates that longer-term contracts are associated with higher customer retention rates.**
<br>
<br>
- **The company could offer new customers longer contracts with rewards, like discounts or extra services, to encourage them to stay longer.**

# Services: 

In [147]:
#Count number of services
def count_services(df):

  num_services = 0
  for service in ['MultipleLines', 'OnlineSecurity', 'OnlineBackup',
               'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies']:
    if df[service] == 'Yes':
      num_services += 1
  return num_services


df['NumOfServices'] = df.apply(count_services, axis=1)
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... StreamingTV  \
0  No phone service             DSL             No  ...          No   
1                No             DSL            Yes  ...          No   
2                No             DSL            Yes  ...          No   
3  No phone service             DSL            Yes  ...          No   
4                No     Fiber optic             No  ...          No   

  StreamingMovies        Contract PaperlessBilling              PaymentMethod  \
0              No  Month-to-month              Yes           Electronic check   
1              No        One year               No               Mailed check   
2              No  Month-to-month              Yes               Mailed check   
3              No        One year               No  Bank transfer (automatic)   
4              No  Month-to-month              Yes           Electronic check   

  MonthlyCharges TotalCharges Churn      tenure_category  NumOfServices  
0          29.85        29.85    No         Short Period              1  
1          56.95      1889.50    No  Intermediate Period              2  
2          53.85       108.15   Yes         Short Period              2  
3          42.30      1840.75    No          Long Period              3  
4          70.70       151.65   Yes         Short Period              0  

[5 rows x 23 columns]

In [148]:
df.groupby('NumOfServices')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

NumOfServices Churn  percentage
0               0    No   79.244151
1               0   Yes   20.755849
2               1    No   65.544041
3               1   Yes   34.455959
4               2    No   62.277952
5               2   Yes   37.722048
6               3    No   68.711656
7               3   Yes   31.288344
8               4    No   74.276527
9               4   Yes   25.723473
10              5    No   78.254848
11              5   Yes   21.745152
12              6    No   88.333333
13              6   Yes   11.666667
14              7    No   94.711538
15              7   Yes    5.288462

- **Customers who subscribe to more services tend to have lower churn rates. For instance, customers with 7 services have a significantly lower churn rate compared to customers with fewer services.**
- **This suggests that bundling multiple services might lead to better customer keeping.**
<br>
<br>
- **The data indicates that offering bundled services could be a potential strategy for reducing churn. Encouraging customers to subscribe to multiple services might increase their commitment to the company.**
- **Consider marketing campaigns that promote the benefits of subscribing to a comprehensive package of services.**

## InternetService

In [149]:
df.groupby('InternetService')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

InternetService Churn  percentage
0             DSL    No   81.040892
1             DSL   Yes   18.959108
2     Fiber optic    No   58.107235
3     Fiber optic   Yes   41.892765
4              No    No   92.595020
5              No   Yes    7.404980

- **Customers with fiber optic internet service seem to have a significantly higher churn rate compared to those with DSL or no internet service.**
- **Customers without any internet service appear to have the lowest churn rate.**
<br>
<br>
- **The type of internet service subscribed to seems to be influencing churn rates.**
- **The relatively higher churn rate among fiber optic customers might indicate that there are issues or factors related to this type of service that are causing dissatisfaction.**

In [150]:
df[['InternetService','tenure','MonthlyCharges','TotalCharges']].groupby('InternetService').mean().round(1).reset_index()

InternetService  tenure  MonthlyCharges  TotalCharges
0             DSL    32.8            58.1        2120.1
1     Fiber optic    32.9            91.5        3205.3
2              No    30.5            21.1         671.6

- **Customers with Fiber optic internet service have noticeably higher average monthly charges (around 91.5 dollar) compared to customers with DSL internet service (around 58.1 dollar) and those with no internet service (around 21.1 dollar).**
- **The difference in average monthly charges between Fiber optic and DSL customers could potentially contribute to the higher churn rate among Fiber optic customers.**

In [151]:
# Group by the internet service type and the customer's contract type
df.groupby('InternetService')['Contract'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

InternetService        Contract  percentage
0             DSL  Month-to-month   50.516316
1             DSL        Two year   25.939694
2             DSL        One year   23.543990
3     Fiber optic  Month-to-month   68.733850
4     Fiber optic        One year   17.409561
5     Fiber optic        Two year   13.856589
6              No        Two year   41.808650
7              No  Month-to-month   34.338139
8              No        One year   23.853211

- **We can say that the high rate of churn for fiber optic customers is due to the fact that they are more likely to have month-to-month contracts. Which give them more flexibility to switch to a competitor if they are unhappy with the service. This is something that the company should consider when trying to reduce churn rates.**

In [152]:
df.groupby('NumOfServices')['InternetService'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

NumOfServices InternetService  percentage
0               0              No   71.025795
1               0             DSL   14.997001
2               0     Fiber optic   13.977205
3               1     Fiber optic   37.219344
4               1             DSL   33.246978
5               1              No   29.533679
6               2     Fiber optic   54.127482
7               2             DSL   45.872518
8               3     Fiber optic   56.850716
9               3             DSL   43.149284
10              4     Fiber optic   60.557342
11              4             DSL   39.442658
12              5     Fiber optic   59.556787
13              5             DSL   40.443213
14              6     Fiber optic   57.619048
15              6             DSL   42.380952
16              7     Fiber optic   58.173077
17              7             DSL   41.826923

- **It seems that the number of services a customer subscribes to does not directly explain why the churn rate for fiber optics is higher. This conclusion is drawn from the observation that fiber optic internet service has a higher number of subscriptions across various numbers of services, yet it still exhibits a higher churn rate.**

- **In essence, there might be factors other than the number of services that are contributing to the higher churn rate for fiber optic customers.**

In [153]:
df.groupby('OnlineSecurity')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

OnlineSecurity Churn  percentage
0                   No    No   58.233276
1                   No   Yes   41.766724
2  No internet service    No   92.595020
3  No internet service   Yes    7.404980
4                  Yes    No   85.388806
5                  Yes   Yes   14.611194

- **Customers with online security are less likely to churn than customers without online security.**
- **There are a few possible explanations for this difference. One possibility is that customers with online security are more satisfied with their service. Another possibility is that customers with online security are more likely to be aware of the risks of not having online security and are therefore less likely to cancel their service.**

In [154]:
df.groupby('OnlineBackup')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

OnlineBackup Churn  percentage
0                   No    No   60.071244
1                   No   Yes   39.928756
2  No internet service    No   92.595020
3  No internet service   Yes    7.404980
4                  Yes    No   78.468506
5                  Yes   Yes   21.531494

- **Customers with online backup may be more satisfied with their service and therefore less likely to churn.**
- **Customers with no internet service may be less likely to need online backup and therefore more likely to churn**

In [155]:
df.groupby('DeviceProtection')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

DeviceProtection Churn  percentage
0                   No    No   60.872375
1                   No   Yes   39.127625
2  No internet service    No   92.595020
3  No internet service   Yes    7.404980
4                  Yes    No   77.497936
5                  Yes   Yes   22.502064

- **The data shows that customers with device protection are less likely to churn than customers without device protection.**
- **Customers with device protection may be more satisfied with their service and therefore less likely to churn.**

In [156]:
# Create a new column representing the combination of three services
# OnlineSecurity', 'OnlineBackup', and 'DeviceProtection'
df['ThreeServicesCombination'] = df['OnlineSecurity'] + data['OnlineBackup'] + data['DeviceProtection']

# Filter data for the desired combination of three services
desired_combination = 'YesYesYes' 
filtered_data = data[data['ThreeServicesCombination'] == desired_combination]

# Analyze churn and other factors for the filtered data
churn_analysis = filtered_data['Churn'].value_counts(normalize=True) * 100
print("Churn Analysis for the Three Services Combination:")
print(churn_analysis)

# analyzing Contract type for the specific combination
contract_analysis = filtered_data['Contract'].value_counts(normalize=True) * 100
print("\nContract Analysis for the Three Services Combination:")
print(contract_analysis)


Churn Analysis for the Three Services Combination:
No     92.063492
Yes     7.936508
Name: Churn, dtype: float64

Contract Analysis for the Three Services Combination:
Two year          61.327561
One year          24.531025
Month-to-month    14.141414
Name: Contract, dtype: float64


- **The majority of customers who have subscribed to the specific combination of three services have opted for longer-term contracts.**
- **Only about 14.14% have selected a 'Month-to-month' contract, which is generally associated with higher churn rates.**
- **The combination of these three services seems to attract customers who are interested in longer-term contracts. This can be attributed to the nature of the services that promote enhanced security and protection.**
- **The low churn rate further supports the idea that customers who value these services are more likely to stay with the company.**
- **Offer this 3 services as a package or give discount.**

In [157]:
df.groupby('StreamingTV')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

StreamingTV Churn  percentage
0                   No    No   66.476868
1                   No   Yes   33.523132
2  No internet service    No   92.595020
3  No internet service   Yes    7.404980
4                  Yes    No   69.929812
5                  Yes   Yes   30.070188

- **The churn rate for customers with StreamingTV ('Yes') is slightly higher at 30.07% compared to those without ('No') at 33.52%.**
- **Customers without any internet service ('No internet service') continue to exhibit the lowest churn rate.**
- **Subscribing to StreamingTV does not seem to have a substantial impact on churn rates, as the difference in churn rates between customers with and without StreamingTV is relatively small.**

In [158]:
# I want to see the correlation between StreamingTV and StreamingMovies

# Calculate the probability of subscribing to StreamingTV
prob_streaming_tv = (df['StreamingTV'] == 'Yes').mean() #calculate the mean of boolean series 

# Calculate the probability of subscribing to both StreamingTV and StreamingMovies
prob_streaming_both = ((df['StreamingTV'] == 'Yes') & (df['StreamingMovies'] == 'Yes')).mean()

# Calculate the conditional probability of subscribing to StreamingMovies given StreamingTV
prob_streaming_movies_tv = prob_streaming_both / prob_streaming_tv

print("Probability of subscribing to StreamingTV:", prob_streaming_tv)
print("Probability of subscribing to both StreamingTV and StreamingMovies:", prob_streaming_both)
print("Conditional probability of subscribing to StreamingMovies given StreamingTV:", prob_streaming_movies_tv)

Probability of subscribing to StreamingTV: 0.38435325855459324
Probability of subscribing to both StreamingTV and StreamingMovies: 0.2754508022149652
Conditional probability of subscribing to StreamingMovies given StreamingTV: 0.7166605097894347


- **The conditional probability of a customer subscribing to StreamingMovies given that they have already subscribed to StreamingTV is approximately 0.717.**
- **There is strong positive relationship between subscribing to StreamingTV and subscribing to StreamingMovies.**
- **The high conditional probability suggests that there's a notable tendency for customers who have chosen to subscribe to StreamingTV to also choose to subscribe to StreamingMovies.**
- **Companies can consider bundling or promoting these services together as a package.**

## Customer Demographics:

In [159]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... StreamingMovies  \
0  No phone service             DSL             No  ...              No   
1                No             DSL            Yes  ...              No   
2                No             DSL            Yes  ...              No   
3  No phone service             DSL            Yes  ...              No   
4                No     Fiber optic             No  ...              No   

         Contract PaperlessBilling              PaymentMethod MonthlyCharges  \
0  Month-to-month              Yes           Electronic check          29.85   
1        One year               No               Mailed check          56.95   
2  Month-to-month              Yes               Mailed check          53.85   
3        One year               No  Bank transfer (automatic)          42.30   
4  Month-to-month              Yes           Electronic check          70.70   

  TotalCharges Churn      tenure_category  NumOfServices  \
0        29.85    No         Short Period              1   
1      1889.50    No  Intermediate Period              2   
2       108.15   Yes         Short Period              2   
3      1840.75    No          Long Period              3   
4       151.65   Yes         Short Period              0   

   ThreeServicesCombination  
0                   NoYesNo  
1                  YesNoYes  
2                  YesYesNo  
3                  YesNoYes  
4                    NoNoNo  

[5 rows x 24 columns]

In [160]:
df.groupby('gender')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

gender Churn  percentage
0  Female    No   73.079128
1  Female   Yes   26.920872
2    Male    No   73.839662
3    Male   Yes   26.160338

- **Both females and males have similar churn rates.**
- **Gender does not appear to be a significant factor in analyzing churn rate.**

In [161]:
df.groupby('SeniorCitizen')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

SeniorCitizen Churn  percentage
0              0    No   76.393832
1              0   Yes   23.606168
2              1    No   58.318739
3              1   Yes   41.681261

- **Senior citizens (value of 1) have a higher churn rate compared to non-senior citizens (value of 0).**
- **Being a senior citizen seems to be associated with a higher rate of churning.**
- **We should consider to find out if there are any services, pricing plans, or customer service options that can be changed to better serve senior customers.**

In [162]:
df.groupby('Partner')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

Partner Churn  percentage
0      No    No   67.042021
1      No   Yes   32.957979
2     Yes    No   80.335097
3     Yes   Yes   19.664903

- **Having a partner appears to be associated with a lower churn rate.**

In [163]:
df.groupby('Dependents')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

Dependents Churn  percentage
0         No    No   68.720860
1         No   Yes   31.279140
2        Yes    No   84.549763
3        Yes   Yes   15.450237

- **Having dependents appears to be associated with a lower churn rate.**

In [164]:
#  select only non-senior citizens with a partner and dependents

df_selected = df.loc[(df['SeniorCitizen'] == 0) &(df['Partner'] == 'Yes') &(df['Dependents'] == 'Yes')]

In [165]:
df_selected.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
10  9763-GRSKD    Male              0     Yes        Yes      13          Yes   
15  3655-SNQYZ  Female              0     Yes        Yes      69          Yes   
18  4190-MFLUW  Female              0     Yes        Yes      10          Yes   
24  6322-HRPFA    Male              0     Yes        Yes      49          Yes   
26  6467-CHFZW    Male              0     Yes        Yes      47          Yes   

   MultipleLines InternetService OnlineSecurity  ... StreamingMovies  \
10            No             DSL            Yes  ...              No   
15           Yes     Fiber optic            Yes  ...             Yes   
18            No             DSL             No  ...              No   
24            No             DSL            Yes  ...              No   
26           Yes     Fiber optic             No  ...             Yes   

          Contract PaperlessBilling            PaymentMethod MonthlyCharges  \
10  Month-to-month              Yes             Mailed check          49.95   
15        Two year               No  Credit card (automatic)         113.25   
18  Month-to-month               No  Credit card (automatic)          55.20   
24  Month-to-month               No  Credit card (automatic)          59.60   
26  Month-to-month              Yes         Electronic check          99.35   

   TotalCharges Churn      tenure_category  NumOfServices  \
10       587.45    No  Intermediate Period              1   
15      7895.15    No          Long Period              7   
18       528.35   Yes         Short Period              2   
24      2970.30    No          Long Period              3   
26      4749.15   Yes          Long Period              4   

    ThreeServicesCombination  
10                   YesNoNo  
15                 YesYesYes  
18                   NoNoYes  
24                  YesYesNo  
26                   NoYesNo  

[5 rows x 24 columns]

In [166]:
df_selected.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1666 entries, 10 to 7040
Data columns (total 24 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   customerID                1666 non-null   object 
 1   gender                    1666 non-null   object 
 2   SeniorCitizen             1666 non-null   int64  
 3   Partner                   1666 non-null   object 
 4   Dependents                1666 non-null   object 
 5   tenure                    1666 non-null   int64  
 6   PhoneService              1666 non-null   object 
 7   MultipleLines             1666 non-null   object 
 8   InternetService           1666 non-null   object 
 9   OnlineSecurity            1666 non-null   object 
 10  OnlineBackup              1666 non-null   object 
 11  DeviceProtection          1666 non-null   object 
 12  TechSupport               1666 non-null   object 
 13  StreamingTV               1666 non-null   object 
 14  Streami

In [167]:
df_selected['Churn'].value_counts(normalize=True) * 100

No     86.254502
Yes    13.745498
Name: Churn, dtype: float64

- **The churn rate for customers who are not senior citizens, have partners, and have dependents is 13.75%, while the churn rate for all customers is 26.54%. This means that customers who fall into this category are less likely to churn.**

## Monthly Charges and Total Charges:

In [171]:
# Calculate average monthly charges for churned and non-churned customers
average_monthly_charges_churned = df[df['Churn'] == 'Yes']['MonthlyCharges'].mean()
average_monthly_charges_not_churned = df[df['Churn'] == 'No']['MonthlyCharges'].mean()

# Calculate total charges for churned and non-churned customers
average_total_charges_churned = df[df['Churn'] == 'Yes']['TotalCharges'].mean()
average_total_charges_not_churned = df[df['Churn'] == 'No']['TotalCharges'].mean()

# Print the results
print('Average monthly charges:')
print('Churn:', average_monthly_charges_churned)
print('Not churn:', average_monthly_charges_not_churned)

print('\nAverage Total charges:')
print('Churn:', average_total_charges_churned )
print('Not churn:', average_total_charges_not_churned)

Average monthly charges:
Churn: 74.44133226324237
Not churn: 61.26512369540008

Average Total charges:
Churn: 1531.7960941680042
Not churn: 2554.7657721007463


- **The average monthly charges for churned customers are higher than the average monthly charges for non-churned customers. This means that churned customers are more likely to have higher monthly bills.**
- **The average total charges for churned customers are also higher than the average total charges for non-churned customers. This means that churned customers are more likely to have spent more money on the service over time.**
- **The difference in average monthly charges is not very significant. However, the difference in average total charges is more significant. This suggests that churned customers are more likely to have been with the company for longer periods of time.**

In [172]:
df.groupby('PaperlessBilling')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

PaperlessBilling Churn  percentage
0               No    No   83.669916
1               No   Yes   16.330084
2              Yes    No   66.434908
3              Yes   Yes   33.565092

- **Customers who have paperless billing ('Yes') have a higher churn rate of around 33.57%.**
- **Customers with paperless billing have a notably higher churn rate compared to those without paperless billing.**
- **Paperless billing might influence customer behavior and satisfaction, leading to differences in churn rates.**
- **Implement strategies to improve the customer experience for those with paperless billing, potentially reducing churn.**

In [175]:
df.groupby('PaymentMethod')['Churn'].value_counts(normalize=True).mul(100).rename('percentage').reset_index()

PaymentMethod Churn  percentage
0  Bank transfer (automatic)    No   83.290155
1  Bank transfer (automatic)   Yes   16.709845
2    Credit card (automatic)    No   84.756899
3    Credit card (automatic)   Yes   15.243101
4           Electronic check    No   54.714588
5           Electronic check   Yes   45.285412
6               Mailed check    No   80.893300
7               Mailed check   Yes   19.106700

- **Payment method seems to have a significant impact on churn rates.**
- **"Electronic check" users have notably higher churn rates compared to other payment methods.**
- **Customers who pay with electronic checks may face challenges or issues that contribute to their higher churn rate.**
- **Consider encouraging the use of payment methods with lower churn rates.**